### ISSUE TITLE
Multi-line prompts

### ISSUE THREAD

**Issue by tailhook** [kkawakam/rustyline#311]
I want to extend multi-line functionality of the library to have different prompts for different lines, like this:
```
sql> SELECT x
...> FROM y;
```
Or this:
```
zsh> python -c '
quote> print "hello"
quote> '
```

This work two complex things:
1. Make prompt a trait rather than string and pass `dyn Prompt`
2. Break highlighting between lines

The questions are:
1. Should we move prompt logic to `Helper` because it might share a tokenizer and other things with a highlighter and other parts of `Helper` trait?
2. What is a good interface for the trait, I would go with something like:
    ```
    fn get_prompt(&self, prompt_info: PromptInfo, context: Context)
    ```
    To abstract the line number, slice, whole buffer and all other things we might add in the future. But it seems the current code prefers adding that arguments explicitly on the function signature. Is that style important?
3. And how to cope with highlighting:
   * We can try to parse all the highlighting codes, reset the highlighting at the end of a line and reinsert all of them at the start of the next line after prompt
   * We can require a highlighter to close highlight line by line

    Both have some trade-offs, I'm not sure which one to choose.

Any other hints are also appreciated.

**Comment by gwenn** [comment]
Ok,
You want [this](https://python-prompt-toolkit.readthedocs.io/en/master/pages/asking_for_input.html#multiline-input).

(a) You can already do it manually in your code like sqlite [shell](https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L625) but it's tedious and user can't fix previous incomplete line(s) (except with your [trick](https://github.com/kkawakam/rustyline/pull/231#issuecomment-570570104)).

(b) If you can do without the continuation prompt, something like this:
```
sql> SELECT x
FROM y;
```
```
zsh> python -c '
print "hello"
'
```
you can try something like this (not tested):
```rust
    rl.bind_sequence(KeyPress::Enter, Cmd::SelfInsert(1, '\n'));
    rl.bind_sequence(KeyPress::Meta('\r'), Cmd::AcceptLine);
```
but user must know when input is incomplete/complete.

(c) If you can do without the continuation prompt, but you want to detect in your code when the input is incomplete/complete, we can try to fix PR #231 such as input cannot be modified while validating but still provide an API which will make possible to do such mutation in the future.

**Comment by tailhook** [comment]
> we can try to fix PR #231 such as input cannot be modified while validating but still provide an API which will make possible to do such mutation in the future.

Yes, let's do that. I'll bring my idea on that PR.

----

But let's also add a continuation prompt too. It isn't that hard per se. And I will do all the work. Let's just agree on API surface so I wouldn't need to rewrite it several times.

**Comment by gwenn** [comment]
Partially fixed in version 6.0.0.
(no continuation prompt)
